### ISSUE TITLE
Incorrect error of use of unassigned variable

### ISSUE THREAD

**Issue by NN---** [Microsoft/TypeScript#13811]
**TypeScript Version:**  2.1.5
strictNullChecks: true

**Code**

```ts
let a: {};

if (!a) { // <!-- Error
   ....
}
```

**Expected behavior:**
Code compiles fine.

**Actual behavior:**
Use of unassigned variable.

**Comment by devdoomari** [comment]
a has the type of {}, and has no value assigned.

so what you wanted:
```
let a = {}; // not let a: {}
```

**Comment by NN---** [comment]
@devdoomari It is not what I want.
I want to have it undefined and assign it only if it is undefined.

Full code:

```ts
let a: {};

for (const x of y) {
  if(x.q) {
    a =x.q;
    break;
  }
}

// Assign default value if couldn't assign before
if (!a) {
  a= {};
}
```

**Comment by ahejlsberg** [comment]
The control flow checker (in strict null checks mode) assures that you'll never accidentally observe `undefined` in a variable unless the type of the variable allows it. Your program logic is constructed such that `undefined` is actually a permitted value (at least in parts of the logic), so I'd recommend including `undefined` in the type of `a`:

```ts
let a: {} | undefined;
```

The problem with your suggestion is that we'd no longer catch this:

```ts
let a: string;
// Intended to have assignment here, but forgot
if (!a) {
    // Think I have an empty string, but I have undefined
}
```

**Comment by NN---** [comment]
@ahejlsberg What about this case ?

I would expect it compiling fine since I am checking against undefined.
```ts
let a: number;
if (typeof a !== "undefined") {
	a = 1;
}
```

**Comment by ahejlsberg** [comment]
I think the same argument holds. If the declared type of a variable does not include `undefined`, the control flow analyzer takes it to mean you never want to observe the variable in an uninitialized state (in any way). I think the consistency of that is important, particularly since the type `number | undefined` in cases where you want to do those types of checks.

**Comment by NN---** [comment]
That's what I ended with.
Thanks.

**Comment by NN---** [comment]
@ahejlsberg 
I feel that there is a need for language construct or helper which tests for null or undefined.
Currently I use "if(a) {..}" but if type of 'a' changes to string or number, this test will not be correct.

I feel more safe using this helper now.
```ts
function isNullOrUndefined<T>(obj: T | null | undefined): obj is null | undefined {
	return typeof obj === "undefined" || obj === null;
}

if (!isNullOrUndefined(a)) { ... }
```

But prefer to have something in the language like:
```ts
iff (a) { ... }
```

**Comment by ahejlsberg** [comment]
Adding entirely new statements beyond what JavaScript has is definitely out of scope for TypeScript. To check if a value is `null` *or* `undefined` in JavaScript, just use `x == null`. See more discussion [here](http://stackoverflow.com/questions/2559318/how-to-check-for-an-undefined-or-null-variable-in-javascript).

**Comment by NN---** [comment]
Thanks.
Didn't think about this pattern
